core/modules/parser/QSMySqlListener.cc

```diff
@@ -1128,74 +1222,81 @@ class PredicateExpressionAdapter :
-        _prepBoolFactor();
-        _boolFactor->addBoolFactorTerm(inPredicate);
+        _getBoolFactor()->addBoolFactorTerm(inPredicate);
     }
 
     void handleExpressionAtomPredicate(shared_ptr<query::ValueExpr> const & valueExpr,
             antlr4::ParserRuleContext* childCtx) override {
+        TRACE_CALLBACK_INFO(valueExpr);
         _prepValueExpr();
         _valueExpr = valueExpr;
     }
 
-    void handleExpressionAtomPredicate(shared_ptr<query::BoolFactorTerm> const & boolFactorTerm,
+    void handleExpressionAtomPredicate(shared_ptr<query::BoolTerm> const & boolTerm,
             antlr4::ParserRuleContext* childCtx) override {
-        _prepBoolFactor();
-        _boolFactor->addBoolFactorTerm(boolFactorTerm);
+        TRACE_CALLBACK_INFO(boolTerm);
+        ASSERT_EXECUTION_CONDITION(nullptr == _boolTerm && nullptr == _valueExpr, "unexpected", _ctx);
+        _boolTerm = boolTerm;
     }
 
     void handleLikePredicate(shared_ptr<query::LikePredicate> const & likePredicate) override {
-        _prepBoolFactor();
-        _boolFactor->addBoolFactorTerm(likePredicate);
+        _getBoolFactor()->addBoolFactorTerm(likePredicate);
+    }
+
+    void handleIsNullPredicate(shared_ptr<query::NullPredicate> const & nullPredicate) override {
+        _getBoolFactor()->addBoolFactorTerm(nullPredicate);
     }
 
     void onExit() {
-        ASSERT_EXECUTION_CONDITION(nullptr != _valueExpr || nullptr != _boolFactor,
+        ASSERT_EXECUTION_CONDITION(nullptr != _valueExpr || nullptr != _boolTerm,
                 "PredicateExpressionAdapter was not populated.", _ctx);
-        if (_boolFactor != nullptr) {
-            lockedParent()->handlePredicateExpression(_boolFactor);
-        } else if (_valueExpr != nullptr) {
-            lockedParent()->handlePredicateExpression(_valueExpr);
+        if (_boolTerm != nullptr) {
+            lockedParent()->handlePredicateExpression(_boolTerm, _ctx);
         } else {
-
+            lockedParent()->handlePredicateExpression(_valueExpr);
         }
     }
 
     string name() const override { return getTypeName(this); }
 
 private:
-    void _prepBoolFactor() {
+    shared_ptr<query::BoolFactor> _getBoolFactor() {
         ASSERT_EXECUTION_CONDITION(nullptr == _valueExpr, "Can't use PredicateExpressionAdapter for " <<
                 "BoolFactor and ValueExpr at the same time.", _ctx);
-        if (nullptr == _boolFactor) {
-            _boolFactor = make_shared<query::BoolFactor>();
+        if (nullptr == _boolTerm) {
+            auto boolFactor = make_shared<query::BoolFactor>();
+            _boolTerm = boolFactor;
+            return boolFactor;
         }
+        auto boolFactor = dynamic_pointer_cast<query::BoolFactor>(_boolTerm);
+        ASSERT_EXECUTION_CONDITION(nullptr != boolFactor, "Can't cast boolTerm:" << _boolTerm <<
+                " to a BoolFactor.", _ctx)
+        return boolFactor;
```

Comment:
Thanks. It *was* convoluted, and looking closer made me realize I was confusing the role of BoolTerm and BoolFactor in this function (though they are similar types they should not be operating in the same context; it's an either-or thing. I added a 3rd member variable to hold the BoolFactor in cases where it's in use, and made it so that it can be set or the BoolTerm (or ValueExpr) can be set, but never can they be set at the same time.

...hopefully that makes a little bit of sense :-)

Reply:
👍